src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs

```diff
@@ -2328,137 +2379,221 @@ private void MeasurePrompts(int startLine, int endLine, out int minPromptLength,
             {
-                if (!TextView.Selection.IsEmpty)
+                using (var transaction = UndoHistory_TestOnly.CreateTransaction(InteractiveWindowResources.Backspace))
                 {
-                    if (CutOrDeleteSelection(isCut: true))
+
+                    if (DeleteHelper(isBackspace: true))
                     {
-                        MoveCaretToClosestEditableBuffer();
+                        // don't record undo histroy for standard-input
+                        if (!ReadingStandardInput)
+                        {
+                            transaction.Complete();
+                        }
                     }
-                    return;
-                }
-
-                var caretPosition = TextView.Caret.Position.BufferPosition;
-
-                // don't cut and move caret if it's in readonly buffer (except in active prompt)
-                if (MapToEditableBuffer(caretPosition) != null ||
-                    IsInActivePrompt(caretPosition))
-                {
-                    CutOrDeleteCurrentLine(isCut: true);
-                    Home(false);
                 }
+                // DeleteHelper handles deleteion completely, 
+                // so always return true to indicate the backspace is done.
+                return true;
             }
 
             /// <summary>Implements <see cref="IInteractiveWindowOperations.Delete"/>.</summary>
             public bool Delete()
             {
                 _historySearch = null;
-                bool handled = false;
-                if (!TextView.Selection.IsEmpty)
+                using (var transaction = UndoHistory_TestOnly.CreateTransaction(InteractiveWindowResources.Delete))
+                {
+                    if (DeleteHelper(isBackspace: false))
+                    {
+                        // don't record undo histroy for standard-input
+                        if (!ReadingStandardInput)
+                        {
+                            transaction.Complete();
+                        }
+                    }
+                }
+                // DeleteHelper handles deleteion completely, 
+                // so always return true to indicate the delete is done.
+                return true;
+            }
+
+            private bool DeleteHelper(bool isBackspace)
+            {
+                var selection = TextView.Selection;
+
+                if (!selection.IsEmpty)
                 {
-                    if (CutOrDeleteSelection(isCut: false))
+                    // do not delete anything if any part of selection is not in current submission
+                    if (!IsSelectionInsideCurrentSubmission())
+                    {
+                        return false;
+                    }
+
+                    if (IsEmptyBoxSelection())
+                    {
+                        return isBackspace ? EditorOperations.Backspace() : EditorOperations.Delete();                        
```

Comment:
This is because I have to duplicate code for deleting empty box selection here otherwise. The reason I use custom method below is to make sure all delete operations will have a consistent behavior (as some of them are not deleting current selection, thus can't handled by `EditorOperations.Delete()`)